lib/util.c

```diff
@@ -344,6 +345,45 @@ msp_binary_interval_search(double query, const double *values, size_t n_values)
     return l;
 }
 
+/* This function follows standard semantics of:
+ *   std::upper_bound() from the standard C++ <algorithm> library
+ *   and numpy.searchsorted(..., side='right') (except numpy NaN query differs)
+ * PRE-CONDITION:
+ *   1) `values` are sorted and not NaN
+ * RETURNS:
+ *   First (leftmost) `index` of strict upper bounds of `query`
+ *   **or** `n_values` if all `values` are lower bounds of `query` or `query` is NaN
```

Comment:
I don't think we should both worrying about NaN semantics here - we can just rule it out on input. We have a pretty tightly circumscribed application here, and I'm sure we'll never need to deal with NaNs.

Reply:
For my next commit I've simplified the function comments to not talk about NaN, except for one caveat at the end that the function does not match the NaN logic of numpy but rather matches std::upper_bound.
Unfortunately, simplifying `!(elements[mid] > query)` to `elements[mid] <= query` means the function will match the semantics of **neither** nump.searchsorted nor std::upper_bound, which seems kind of a pity.

The std::upper_bound/lower_bound semantics for NaN make a ton of sense to me. Granted I probably think about NaN an abnormal amount. :)

Don't write off NaN so fast! It's an elegant and powerful error handling option in the absence of exception handling and error codes. I'm not surprised it exists as the result of `INFINITY * 0` and 'x / 0'. In the future you might find NaN as a useful way to gracefully handling very tricky numerical challenges.